crates/topos-tce-api/src/runtime/sync_task.rs

```diff
@@ -0,0 +1,179 @@
+use topos_core::types::CertificateDelivered;
+use topos_core::uci::SubnetId;
+use topos_tce_storage::{FetchCertificatesFilter, FetchCertificatesPosition, StorageClient};
+use tracing::{error, info};
+use uuid::Uuid;
+
+type TargetSubnetStreamPositions = HashMap<SubnetId, HashMap<SubnetId, TargetStreamPosition>>;
+pub(crate) type RunningTasks =
+    FuturesUnordered<Pin<Box<dyn Future<Output = SyncTaskStatus> + Send>>>;
+
+/// Status of a sync task
+///
+/// When registering a stream, a sync task is started to fetch certificates from the storage
+/// and push them to the stream.
+#[derive(Debug)]
+pub(crate) enum SyncTaskStatus {
+    ///  The sync task is active and started running
+    Running,
+    /// The sync task failed and reported an error
+    Error,
+    /// The sync task exited gracefully and is done pushing certificates to the stream
+    Done,
+    /// The sync task was cancelled externally
+    Cancelled,
+}
+
+pub(crate) struct SyncTask {
+    /// The status of the SyncTask. Can be used to check if the task is still running
+    pub(crate) status: SyncTaskStatus,
+    /// The stream with which the SyncTask is connected and pushes certificates to
+    pub(crate) stream_id: Uuid,
+    /// The positions of each subnet in the stream of where the stream is currently left of
+    pub(crate) target_subnet_stream_positions: TargetSubnetStreamPositions,
+    /// The connection to the database layer through a StorageClient
+    pub(crate) storage: StorageClient,
+    /// The notifier is used to send certificates to the stream
+    pub(crate) notifier: Sender<StreamCommand>,
+    /// If a new stream is registered with the same Uuid, the sync task will be cancelled
+    pub(crate) cancel_token: CancellationToken,
+}
+
+impl SyncTask {
+    /// Creating a new SyncTask which will fetch certificates from the storage and pushes them to the stream
+    pub(crate) fn new(
+        stream_id: Uuid,
+        target_subnet_stream_positions: TargetSubnetStreamPositions,
+        storage: StorageClient,
+        notifier: Sender<StreamCommand>,
+        cancel_token: CancellationToken,
+    ) -> Self {
+        println!("CREATED A NEW TASK");
+        Self {
+            status: SyncTaskStatus::Running,
+            stream_id,
+            target_subnet_stream_positions,
+            storage,
+            notifier,
+            cancel_token,
+        }
+    }
+}
+
+impl IntoFuture for SyncTask {
+    type Output = SyncTaskStatus;
+
+    type IntoFuture = Pin<Box<dyn Future<Output = Self::Output> + Send + 'static>>;
+
+    fn into_future(mut self) -> Self::IntoFuture {
+        Box::pin(async move {
+            println!("START FUTURE");
+            info!("Sync task started for stream {}", self.stream_id);
+            let mut collector: Vec<(CertificateDelivered, FetchCertificatesPosition)> = Vec::new();
+
+            for (target_subnet_id, source) in &mut self.target_subnet_stream_positions {
+                if self.cancel_token.is_cancelled() {
```

Comment:
Why do we check the cancellation status inside the loop? Do we need to check it for every iteration?

Reply:
`async` call inside the loop, so I thought this might take longer, and we want to cancel immedietaly if the same `uuid` stream is registering again.